aptos-move/aptos-vm/src/move_vm_ext/vm.rs

```diff
@@ -63,10 +72,26 @@ impl MoveVmExt {
                     paranoid_type_checks: crate::AptosVM::get_paranoid_checks(),
                 },
             )?,
+            cache: RwLock::new(HashMap::new()),
             chain_id,
         })
     }
 
+    pub fn get_module_metadata(&self, module_id: ModuleId) -> Arc<Option<RuntimeModuleMetadataV1>> {
+        let read = self.cache.read().unwrap();
+        if let Some(ret) = read.get(&module_id) {
+            ret.clone()
+        } else {
+            drop(read);
+            let ret = Arc::new(aptos_framework::get_vm_metadata(
+                &self.inner,
+                module_id.clone(),
+            ));
+            self.cache.write().unwrap().insert(module_id, ret.clone());
+            ret
+        }
```

Comment:
can we verify that a previous write hasn't already done this work?
seems like a better approach might be to get a write lock, check if it doesn't exist, set it if it doesn't, then perform a read -- of course then you have the yuckiness of handling an impossible get that is conditional though it should have been set...

Reply:
I don't think it'd make big difference if I understand what you suggested here, currently VM is not accessed concurrently